libtransmission/peer-mgr.h

```diff
@@ -316,17 +351,48 @@ class tr_peer_info
         return ret;
     }
 
-private:
-    [[nodiscard]] constexpr tr_address const& addr() const noexcept
-    {
-        return socket_address_.address();
-    }
+    // ---
 
-    [[nodiscard]] constexpr auto port() const noexcept
+    // merge a peer info objects into ourself
+    void merge(tr_peer_info const& that) noexcept
     {
-        return socket_address_.port();
+        connection_attempted_at_ = std::max(connection_attempted_at_, that.connection_attempted_at_);
+        connection_changed_at_ = std::max(connection_changed_at_, that.connection_changed_at_);
+        piece_data_at_ = std::max(piece_data_at_, that.piece_data_at_);
+
+        /* no need to merge blocklist since it gets updated elsewhere */
+
+        {
+            auto const conn_this = !is_connectable_ || *is_connectable_;
+            auto const conn_that = !that.is_connectable_ || *that.is_connectable_;
+
+            if (conn_this != conn_that)
+            {
+                is_connectable_.reset();
+            }
+            else
+            {
+                set_connectable(conn_this && conn_that);
```

Comment:
This doesn't seem to be right in the case where neither optional is set -- when that happens, we're calling  `set_connectable(true)` here?

Reply:
I overhauled this part at f5f4465, please take a look again.